[TASK] Create a console command tutorial (#2208)
TYPO3-Documentation/TYPO3CMS-Reference-CoreApi: Documentation/ApiOverview/CommandControllers/Tutorial.rst

console command that creates a new command in an extension of your choice:
:ref:`Create a new console command with "Make" <extension-make-console-command>`.
You can use "Make" to create a console command even if your extension was
created by different means.

This command can be found in the
`Examples extension <https://github.com/TYPO3-Documentation/t3docs-examples>`__.

1. Register the command
-----------------------

Register the command in :file:`Configuration/Services.yaml` by adding the service
definition for your class as tag :yaml:`console.command`:

..  code-block:: yaml
    :caption: EXT:examples/Configuration/Services.yaml
    :linenos:
    :emphasize-lines: 11-16

    services:
      _defaults:
        autowire: true
        autoconfigure: true
        public: false

      T3docs\Examples\:
        resource: '../Classes/*'
        exclude: '../Classes/Domain/Model/*'

      T3docs\Examples\Command\DoSomethingCommand:
        tags:
          - name: console.command
            command: 'examples:dosomething'
            description: 'A command that does nothing and always succeeds.'

..  note::
    Despite using :file:`autoconfigure: true` the commands

>     have to be explicitly defined in :file:`Configuration/Services.yaml`. The
>     description is mandatory.

Author: ```suggestion
    have to be explicitly defined in :file:`Configuration/Services.yaml`. It
    is recommended to always supply a description, Otherwise there is 
    an empty space in the list of commands.
```

2. Create the command class
---------------------------

Create a class called :php:`DoSomethingCommand` extending
:php:`\Symfony\Component\Console\Command\Command`.

..  include:: /CodeSnippets/Tutorials/Command/Classes/DoSomethingCommand.rst.txt

The following two methods should be overridden by your class:

:php:`configure()`
    As the name would suggest, allows to configure the command.
    The method allows to add a help text and/or define arguments and options.

:php:`execute()`

>     Contains the logic when executing the command. Should return
>     :php:`Command::SUCCESS` or :php:`Command::FAILURE`.

Author: ```suggestion
    Contains the logic when executing the command. Must 
    return an integer. It is considered best practice to 
    return the constants
    :php:`Command::SUCCESS` or :php:`Command::FAILURE`.
```

.. include:: /Includes.rst.txt

.. _console-command-tutorial:

========
Tutorial
========

Create a console command from scratch
=====================================

> A console command is always situated in an Extension. If you want to create

Reviewer: ```suggestion
A console command is always situated in an extension. If you want to create
```

one, :ref:`kickstart a custom extension <extension-kickstart>` or use your
sitepackage extension.

Creating a basic command
========================

The extension kickstarter ":ref:`Make <extension-make>`" offers a convenient
console command that creates a new command in an extension of your choice:
:ref:`Create a new console command with "Make" <extension-make-console-command>`.
You can use "Make" to create a console command even if your extension was
created by different means.

This command can be found in the

> `Examples extension <https://github.com/TYPO3-Documentation/t3docs-examples>`.

Reviewer: ```suggestion
`Examples extension <https://github.com/TYPO3-Documentation/t3docs-examples>`__.
```

> 1. Register the command

Reviewer: Can we use `..  rst-class:: bignums-xxl` for the numbered headers?

Author: I think they look cluttered when we have big, structured blocks like these

-----------------------

> Register via DI in :file:`Configuration/Services.yaml` by adding the service

Reviewer: ```suggestion
Register the command in :file:`Configuration/Services.yaml` by adding the service
```

As I understand, DI is injecting dependencies into classes. Here we register a command which is something different.

Author: I think you are true, it was already written like that

definition for your class as tag :yaml:`console.command`:

..  code-block:: yaml

>     :caption: EXT:some_extension/Configuration/Services.yaml

Reviewer: ```suggestion
    :caption: EXT:examples/Configuration/Services.yaml
```

    :linenos:

>     :emphasize-lines: 12-16

Reviewer: ```suggestion
    :emphasize-lines: 11-16
```

Would also highlight the FQCN.

    services:
      _defaults:
        autowire: true
        autoconfigure: true
        public: false

      T3docs\Examples\:
        resource: '../Classes/*'
        exclude: '../Classes/Domain/Model/*'

      T3docs\Examples\Command\DoSomethingCommand:
        tags:
          - name: console.command
            command: 'examples:dosomething'
            description: 'A command that does nothing and always succeeds.'

>             schedulable: true

Reviewer: As it is removed now also in EXT:examples, we should remove it here also. What do you think?

>             schedulable: true

Author: ```suggestion
```

..  note::
    Despite using :file:`autoconfigure: true` the commands
    have to be explicitly defined in :file:`Configuration/Services.yaml`. The

>     description is mandatory.

Reviewer: The description is not mandatory. If it's not available the description will be empty in the list.

Author: done

2. Create the command class
---------------------------

> Create a class called :php:`DoThingsCommand` extending

Reviewer: ```suggestion
Create a class called :php:`DoSomethingCommand` extending
```

:php:`\Symfony\Component\Console\Command\Command`.

..  include:: /CodeSnippets/Tutorials/Command/Classes/DoSomethingCommand.rst.txt

The following two methods should be overridden by your class:

:php:`configure()`
    As the name would suggest, allows to configure the command.

>     The method allows to add a help text and/or define arguments.

Reviewer: ```suggestion
    The method allows to add a help text and/or define arguments and options.
```

:php:`execute()`

>     Contains the logic when executing the command. Should return

Reviewer: ```suggestion
    Contains the logic when executing the command. Must return
```

Author: That an integer must be returned will be told by any IDE, however you should return a constant and not something like "0"

Reviewer: The "must" is tied to the integer explanation from above (0 means success, and so on). If you discard that, you can also discard this.

>     :php:`Command::SUCCESS` or :php:`Command::FAILURE`.

Reviewer: ```suggestion
    an integer: `0` means the command is successful, greater `0` means an error occurred. 
    The :php:`Command class` provides three constants for convenience:
    :php:`Command::SUCCESS`, :php:`Command::FAILURE` and :php:`Command::INVALID`.
```

Author: It is best practice to return the constant, not an integer literal

Reviewer: Yup, but to know the background is always good. But it is a suggestion, feel free to discard it.

Author: I made a suggestion

.. seealso::

   A detailed description and an example can be found in
   `the Symfony Command Documentation <https://symfony.com/doc/current/console.html>`_.

3. Run the command
------------------

The above example can be run via command line:

.. tabs::

   .. group-tab:: Composer-based installation

      .. code-block:: bash

>          vendor/bin/typo3 example:dothings

Reviewer: ```suggestion
         vendor/bin/typo3 examples:dosomething
```

   .. group-tab:: Legacy installation

      .. code-block:: bash

>          typo3/sysext/core/bin/typo3 someextension:dothings

Reviewer: ```suggestion
         typo3/sysext/core/bin/typo3 examples:dosomething
```

The command will return without a message as it does nothing but stating it
succeeded.

..  note::
    If a newly created or changed command is not found, clear the cache:

    .. code-block:: bash

>         vendor/bin/typo3 cache:flush

Reviewer: Maybe here also tabs with legacy installation?

Author: as it is already in the note I think the whole layout becomes cluttered then...

.. code-block:: php
    :caption: EXT:examples/Classes/Command/CreateWizardCommand.php

    use Symfony\Component\Console\Style\SymfonyStyle;

    final class CreateWizardCommand extends Command
    {
        protected function execute(
            InputInterface $input,
            OutputInterface $output
        ): int {
            $io = new SymfonyStyle($input, $output);
            // do some user interaction
            return Command::SUCCESS;
        }
    }

The :php:`$io` variable can then be used to generate output and prompt for
input:

..  include:: /CodeSnippets/Tutorials/Command/Classes/CreateWizardCommandIo.rst.txt

Dependency injection in console commands
========================================

You can use :ref:`dependency injection (DI) <Dependency-Injection>` in console
commands by constructor injection or method injection:

..  include:: /CodeSnippets/Tutorials/Command/Classes/DependencyInjection.rst.txt

Initialize backend user
=======================

A backend user can be initialized with this call inside :php:`execute()` method:

..  code-block:: php
    :caption: EXT:some_extension/Classes/Command/DoBackendRelatedThingsCommand.php

>     :emphasize-lines: 7

Reviewer: ```suggestion
    :emphasize-lines: 9
```

:ref:`Create a new console command with "Make" <extension-make-console-command>`.
You can use "Make" to create a console command even if your extension was
created by different means.

This command can be found in the
`Examples extension <https://github.com/TYPO3-Documentation/t3docs-examples>`__.

1. Register the command
-----------------------

Register the command in :file:`Configuration/Services.yaml` by adding the service
definition for your class as tag :yaml:`console.command`:

..  code-block:: yaml
    :caption: EXT:examples/Configuration/Services.yaml
    :linenos:
    :emphasize-lines: 11-16

    services:
      _defaults:
        autowire: true
        autoconfigure: true
        public: false

      T3docs\Examples\:
        resource: '../Classes/*'
        exclude: '../Classes/Domain/Model/*'

      T3docs\Examples\Command\DoSomethingCommand:
        tags:
          - name: console.command
            command: 'examples:dosomething'
            description: 'A command that does nothing and always succeeds.'

..  note::
    Despite using :file:`autoconfigure: true` the commands
    have to be explicitly defined in :file:`Configuration/Services.yaml`. It

>     is recommended to always supply a description, Otherwise there is 

Reviewer: ```suggestion
    is recommended to always supply a description, otherwise there is 
```
